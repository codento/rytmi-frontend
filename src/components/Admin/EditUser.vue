<template>
  <div>
    <b-card :title="fullName">
      <div>
        <b-form-group>
          <b-form-checkbox
            v-model="isAdmin"
            inline
          >
            Is admin?
          </b-form-checkbox>
          <b-form-checkbox
            v-model="isActive"
            inline
          >
            Is active?
          </b-form-checkbox>
        </b-form-group>
        <b-form-group
          id="roleLabel"
          label-cols-sm="2"
          label="Role:"
          label-for="roleInput"
        >
          <v-select
            id="roleInput"
            v-model="selectedEmployeeRoles"
            :options="employeeRoleList(selectedEmployeeRoles)"
            multiple
          />
        </b-form-group>
        <b-form-group>
          <b-row>
            <b-col sm="3">
              <b-button
                variant="success"
                @click="submit"
              >
                Update
              </b-button>
            </b-col>
            <b-col
              sm="3"
              offset-sm="6"
            >
              <b-button
                variant="danger"
                @click="del"
              >
                Delete
              </b-button>
            </b-col>
          </b-row>
        </b-form-group>
      </div>
    </b-card>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import vSelect from 'vue-select'

export default {
  components: { vSelect },
  props: {
    user: {
      type: Object,
      required: true
    },
    update: {
      type: Function,
      required: true
    },
    delete: {
      type: Function,
      required: true
    }
  },
  data () {
    return {
      isAdmin: this.user.admin,
      isActive: this.user.active,
      profile: null,
      selectedEmployeeRoles: []
    }
  },
  computed: {
    ...mapGetters(['employeeRoles', 'profileById', 'profileByUserId']),
    fullName () {
      const { firstName, lastName } = this.user
      return `${firstName} ${lastName}`
    }
  },
  watch: {
    user (newUser) {
      this.isAdmin = newUser.admin
      this.isActive = newUser.active
      this.profile = this.getActiveEmployeeProfile()
      this.selectedEmployeeRoles = this.employeeRoleList([]).filter(role => this.profile.employeeRoles.includes(role.id))
    },
    selectedEmployeeRoles (newRoles) {
      this.profile.employeeRoles = newRoles.map(role => role.id)
    }
  },
  created () {
    this.profile = this.getActiveEmployeeProfile()
    this.selectedEmployeeRoles = this.employeeRoleList([]).filter(role => this.profile.employeeRoles.includes(role.id))
  },
  methods: {
    employeeRoleList (selectedRoles) {
      const roles = this.employeeRoles.map(item => {
        return {
          label: item.title,
          id: item.id
        }
      })
      return roles.filter(role => !selectedRoles.some(selectedRole => selectedRole.id === role.id))
    },
    submit () {
      this.update({ id: this.user.id, active: this.isActive, admin: this.isAdmin }, this.profile)
    },
    del () {
      const promptMessage = `
        Are you sure you want to delete the user?
        This action cannot be reverted!
        Write DELETE to confirm`
      const deletePrompt = prompt(promptMessage)
      if (deletePrompt === 'DELETE') {
        this.delete()
      }
    },
    getActiveEmployeeProfile () {
      return this.profileByUserId(this.user.id)
    }
  }
}
</script>
