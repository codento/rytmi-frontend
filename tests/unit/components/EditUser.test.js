import Vuex from 'vuex'
import BootstrapVue from 'bootstrap-vue'
import { createLocalVue, mount } from '@vue/test-utils'
import EditUser from '@/components/Admin/EditUser'

describe('EditUser.vue', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.use(BootstrapVue)

  const getters = {
    profileByUserId: () => (arg) => {
      return {
        id: 1,
        userId: 2,
        firstName: 'Bar',
        lastName: 'Foo',
        photoPath: '',
        employeeRoles: [1],
        title: 'software developer',
        email: 'foo.bar@foo.com',
        phone: '1354',
        description: 'fdas'
      }
    },
    employeeRoles: () => {
      return [
        { id: 1, title: 'soft mengineer' },
        { id: 2, title: 'hardcore soft engineer' }
      ]
    },
    absencesByProfileId: () => (id) => {},
    leaves: () => {}
  }

  const actions = {
    fetchAbsencesForProfile: () => (id) => {}
  }

  const propsData = {
    user: {
      id: 1,
      firstName: 'Bar',
      lastName: 'Foo',
      active: false,
      admin: false
    },
    update: jest.fn(),
    delete: jest.fn()
  }

  let wrapper, store

  beforeEach(() => {
    store = new Vuex.Store({ getters, actions })
    wrapper = mount(EditUser, {
      store,
      propsData,
      localVue
    })
  })

  it('Template is correct', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
