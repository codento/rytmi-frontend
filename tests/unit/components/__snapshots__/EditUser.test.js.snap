// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUser.vue Template is correct 1`] = `
<div>
  <div
    class="card"
  >
    <!---->
    <!---->
    <div
      class="card-body"
    >
      <h4
        class="card-title"
      >
        Bar Foo
      </h4>
      <!---->
      <div
        class="card"
      >
        <!---->
        <!---->
        <div
          class="card-body"
        >
          <!---->
          <!---->
          <fieldset
            class="form-group"
            id="__BVID__6"
          >
            <!---->
            <div
              role="group"
              tabindex="-1"
            >
              <div
                class="custom-control custom-control-inline custom-checkbox"
              >
                <input
                  autocomplete="off"
                  class="custom-control-input"
                  id="__BVID__7"
                  type="checkbox"
                  value="true"
                />
                <label
                  class="custom-control-label"
                  for="__BVID__7"
                >
                  
          Is admin?
        
                </label>
              </div>
               
              <div
                class="custom-control custom-control-inline custom-checkbox"
              >
                <input
                  autocomplete="off"
                  class="custom-control-input"
                  id="__BVID__8"
                  type="checkbox"
                  value="true"
                />
                <label
                  class="custom-control-label"
                  for="__BVID__8"
                >
                  
          Is active?
        
                </label>
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
           
          <div
            class="form-row form-group"
            id="roleLabel"
            role="group"
          >
            <label
              class="col-sm-2 col-form-label"
              for="roleInput"
              id="roleLabel__BV_label_"
            >
              Role:
            </label>
            <div
              class="col"
            >
              <div
                class="dropdown v-select searchable"
                dir="auto"
                id="roleInput"
              >
                <div
                  class="dropdown-toggle"
                >
                  <div
                    class="vs__selected-options"
                  >
                    <span
                      class="selected-tag"
                    >
                      
            soft mengineer
           
                      <button
                        aria-label="Remove option"
                        class="close"
                        type="button"
                      >
                        <span
                          aria-hidden="true"
                        >
                          ×
                        </span>
                      </button>
                    </span>
                     
                    <input
                      aria-label="Search for option"
                      autocomplete="off"
                      class="form-control"
                      role="combobox"
                      type="search"
                    />
                  </div>
                   
                  <div
                    class="vs__actions"
                  >
                    <button
                      class="clear"
                      style="display: none;"
                      title="Clear selection"
                      type="button"
                    >
                      <span
                        aria-hidden="true"
                      >
                        ×
                      </span>
                    </button>
                     
                    <i
                      class="open-indicator"
                      role="presentation"
                    />
                     
                    <div
                      class="spinner"
                      style="display: none;"
                    >
                      Loading...
                    </div>
                  </div>
                </div>
                 
                <!---->
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </div>
           
          <fieldset
            class="form-group"
            id="__BVID__12"
          >
            <!---->
            <div
              role="group"
              tabindex="-1"
            >
              <div
                class="row"
              >
                <div
                  class="col-sm-3"
                >
                  <button
                    class="btn btn-primary"
                    name="update"
                    type="button"
                  >
                    
              Update
            
                  </button>
                </div>
                 
                <div
                  class="col-sm-3 offset-sm-6"
                >
                  <button
                    class="btn btn-danger"
                    name="delete"
                    type="button"
                  >
                    
              Delete
            
                  </button>
                </div>
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
        </div>
        <!---->
        <!---->
      </div>
       
      <div
        class="card"
      >
        <!---->
        <!---->
        <div
          class="card-body"
        >
          <!---->
          <!---->
          <fieldset
            class="form-group"
            id="__BVID__13"
          >
            <!---->
            <div
              role="group"
              tabindex="-1"
            >
              <form
                class=""
                id="leave-form"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-12 projects-table col"
                  >
                    <table
                      aria-busy="false"
                      aria-colcount="5"
                      aria-describedby="__BVID__15__caption_"
                      class="table b-table b-table-stacked-sm"
                      id="__BVID__15"
                      role="table"
                    >
                      <caption
                        class="b-table-caption-top"
                        id="__BVID__15__caption_"
                      >
                        <h4>
                          Leaves
                        </h4>
                      </caption>
                      <!---->
                      <thead
                        class=""
                        role="rowgroup"
                      >
                        <!---->
                        <tr
                          role="row"
                        >
                          <th
                            aria-colindex="1"
                            class=""
                            role="columnheader"
                            scope="col"
                          >
                            Description
                          </th>
                          <th
                            aria-colindex="2"
                            class=""
                            role="columnheader"
                            scope="col"
                          >
                            From
                          </th>
                          <th
                            aria-colindex="3"
                            class=""
                            role="columnheader"
                            scope="col"
                          >
                            To
                          </th>
                          <th
                            aria-colindex="4"
                            aria-label="Edit"
                            class=""
                            role="columnheader"
                            scope="col"
                          >
                            
                          </th>
                          <th
                            aria-colindex="5"
                            aria-label="Remove"
                            class=""
                            role="columnheader"
                            scope="col"
                          >
                            
                          </th>
                        </tr>
                      </thead>
                      <!---->
                      <tbody
                        class=""
                        role="rowgroup"
                      >
                        <!---->
                        <!---->
                        <!---->
                      </tbody>
                    </table>
                  </div>
                </div>
                 
                <h4>
                  Add leave
                </h4>
                 
                <div
                  class="row mt-4"
                >
                  <div
                    class="col-sm-6"
                  >
                    <small
                      for="leave-description-input"
                    >
                      Leave type *
                    </small>
                     
                    <fieldset
                      class="form-group"
                      id="__BVID__16"
                    >
                      <!---->
                      <div
                        role="group"
                        tabindex="-1"
                      >
                        <!---->
                        <div
                          aria-atomic="true"
                          aria-live="assertive"
                          class="invalid-feedback"
                          role="alert"
                          tabindex="-1"
                        >
                          Required
                        </div>
                        <!---->
                        <!---->
                      </div>
                    </fieldset>
                  </div>
                </div>
                 
                <!---->
                 
                <!---->
                 
                <button
                  class="btn mt-2 btn-primary disabled"
                  disabled="disabled"
                  name="submit"
                  type="submit"
                >
                  
    Add
  
                </button>
                 
                <!---->
              </form>
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
